readme.md
# PyMonitr, a simple system monitor for linux

using [psutil](https://pypi.org/project/psutil/) and [colorhex](https://pypi.org/project/colorhex/)

With PyMonitr you can:
1. **Monitor your RAM, CPU, storage ('/', removed in v9) & swap usage,**
2. **and get information about your CPU, such as:**
+ **Usage,**
+ **Temperature,**
+ **Total logical & non logical cores,**
+ **Architecture type,**
+ **CPU vendor &**
+ **CPU model**

# screenshots:
![screenshot 2](screensh1.png)

# requirements
**[psutil](https://pypi.org/project/psutil/), [colorhex](https://pypi.org/project/colorhex/), [cursor](https://pypi.org/project/cursor)**

**pip3 install psutil colorhex cursor**

# usage
`./pymn_v9.py -OPTIONS`

available options: -r (refresh time). default value: 1

# note
im tired working for this, ive worked so hard and for a long time, but i didnt made it stable, i can fix it, but i dont want to. 9 is the last version of pymonitr, which was made MONTHS ago, i just decided to upload it now, and from this day, i stop updating pymonitr. please use [sysmon](https://github.com/devlocalhost/sysmon) instead, a better version of pymonitr

changelog.md
# Changelog

#### Version: 1
1. **Added colors,**
2. **Added: Cores per cluster, Threads per core, Architecture type**

#### Version: 2
1. **Colors now match the usage. For example, if CPU usage is between 0 between 50 the color will be green, 51 - 80 yellow, and 81+ red and the text will flash too,**
2. **Changed start animation logo speed from 2 to 1**

#### Version: 3
1. **Changed colors of:**
	**+ Total cores,**
	**+ Clusters,**
	**+ Cores per cluster,**
	**+ Threads per core,**
	**+ Core * frequency,**
	**+ Architecture type and CPU model/name to white**

#### Version: 4
1. **Fixed the annoying "minus (-)" thing, where it would usually show up on the "Free" fields (such as free storage etc),**
2. **Changed colors of:**
	**+ Total cores,**
	**+ Clusters,**
	**+ Cores per cluster,**
	**+ Threads per core,**
	**+ Core * frequency,**
	**+ Architecture type and CPU model/name to some random ones**

#### Version: 5
1. **Removed cluster field**

#### Version: 6
1. **Changed & added new colors,**
2. **Removed core frequency field,**
3. **Before the status colors were:**
	**+ 0-50 green,**
	**+ 50-80 yellow,**
	**+ 80+ red,**
	**now its:**
	**+ 0-50 green,**
	**+ 50-80 yellow,**
	**+ 80-90 orange & 90+ red,**
4. **Removed dot animation**

#### Version: 7
1. **Added the version while starting the program,**
2. **Added CPU temperature, which MIGHT NOT be accurate, because im using a laptop that has 1 core (yes, 1 core, im not joking...). The program also reads the "/sys/class/thermal/thermal_zone0/temp" file to get the temperature. If you have any better idea to do this, you can open a issue or something *idk lol***

#### Version: 8
1. **Added CPU vendor & new color,**
2. **Added arguments/flags to control refresh time for the stats,**
3. **Added a exception in case CPU temperature cannot be read,**
4. **The program now will ask you to install the required packages incase you havent installed them, which are: [psutil](https://pypi.org/project/psutil/) and [colorhex](https://pypi.org/project/colorhex/)**

#### Version: 8.3
1. **Added back the CPU frequencies. Now you can see each cores frequency**
2. **Added apps version at the top**
3. **Changed "Total cores" to:**
	**+ Total non logical cores &**
	**+ Total logical cores**
4. **New colors to some fields**

#### Version 9. final version
1. **Network related stats**

pymn_v3.py
from psutil import virtual_memory, cpu_freq, disk_usage, cpu_percent, cpu_count, swap_memory
from subprocess import check_output
from termcolor import colored
from time import sleep
from sys import exit
from os import system

CPU_MODEL = check_output("lscpu | grep 'Model name' | cut -f 2 -d ':' | awk '{$1=$1}1' | tr '\n', ' '", shell=True, text=True)
ARCH_TYPE = check_output("uname -m | tr '\n', ' '", shell=True, text=True)
CLUSTERS = check_output("lscpu | grep 'Socket(s)' | cut -f 2 -d ':' | awk '{$1=$1}1' | tr '\n', ' '", shell=True, text=True)
CORES_PER_CLUSTER = check_output("lscpu | grep 'Core(s) per socket' | cut -f 2 -d ':' | awk '{$1=$1}1' | tr '\n', ' '", shell=True, text=True)
THREADS_PER_CORE = check_output("lscpu | grep 'Thread(s) per core' | cut -f 2 -d ':' | awk '{$1=$1}1' | tr '\n', ' '", shell=True, text=True)


def readable_format(fsize, units=[' bytes',' KB',' MB',' GB',' TB', ' PB', ' EB']):
	return "{:.2f}{}".format(float(fsize), units[0]) if fsize < 1024 else readable_format(fsize / 1024, units[1:])

system('clear')
print(colored(' -- PyMonitr -- ', 'blue', attrs=['bold']))
sleep(1)
system('clear')

print('.')
sleep(0.15)
system('clear')
print('..')
sleep(0.15)
system('clear')
print('...')
sleep(0.15)
system('clear')

while True:
	ram = virtual_memory()
	cpu = cpu_freq(percpu=True)
	storage = disk_usage('/')
	swap = swap_memory()

	total_ram = readable_format(ram[0])
	used_ram = ram[0] - ram[1]
	free_ram = readable_format(ram[1])

	percent_used_ram = ram.percent
	percent_free_ram = abs(percent_used_ram)

	total_swap = readable_format(swap[0])
	used_swap = readable_format(swap[1])
	free_swap = readable_format(swap[2])

	percent_used_swap = swap.percent
	percent_free_swap = abs(percent_used_swap)

	total_storage = readable_format(storage[0])
	used_storage = readable_format(storage[1])
	free_storage = readable_format(storage[2])

	percent_used_storage = storage.percent
	percent_free_storage = abs(percent_used_storage)

	print(colored('|===== [Storage] =====|', 'green', attrs=['bold']))
	print(colored(f'Total: {total_storage}', 'blue', attrs=['bold']))

	if int(percent_used_storage) in range(0, 50):
		print(colored(f'Used: {used_storage} | {percent_used_storage}%', 'green', attrs=['bold']))

	elif int(percent_used_storage) in range(51, 80):
		print(colored(f'Used: {used_storage} | {percent_used_storage}%', 'yellow', attrs=['bold']))

	elif int(percent_used_storage) in range(81, 90):
		print(colored(f'Used: {used_storage} | {percent_used_storage}%', 'red', attrs=['bold']))

	elif int(percent_used_storage) in range(91, 101):
		print(colored(f'Used: {used_storage} | {percent_used_storage}% | STORAGE USAGE IS TOO HIGH!!', 'magenta', attrs=['bold', 'blink']))

	if int(percent_free_storage) in range(0, 50):
		print(colored(f'Free: {free_storage} | {percent_free_storage - 100:.1f}%', 'green', attrs=['bold']))

	elif int(percent_free_storage) in range(51, 80):
		print(colored(f'Free: {free_storage} | {percent_free_storage - 100:.1f}%', 'yellow', attrs=['bold']))

	elif int(percent_free_storage) in range(81, 90):
		print(colored(f'Free: {free_storage} | {percent_free_storage - 100:.1f}%', 'red', attrs=['bold']))

	elif int(percent_free_storage) in range(91, 101):
		print(colored(f'Free: {free_storage} | {percent_free_storage - 100:.1f}% | LOW STORAGE SPACE AVAILABLE!!', 'magenta', attrs=['bold', 'blink']))

	print(colored('|======= [RAM] =======|', 'green', attrs=['bold']))
	print(colored(f'Total: {total_ram}', 'blue', attrs=['bold']))

	if int(percent_used_ram) in range(0, 50):
		print(colored(f'Used: {readable_format(used_ram)} | {percent_used_ram}%', 'green', attrs=['bold']))

	elif int(percent_used_ram) in range(51, 80):
		print(colored(f'Used: {readable_format(used_ram)} | {percent_used_ram}%', 'yellow', attrs=['bold']))

	elif int(percent_used_ram) in range(81, 90):
		print(colored(f'Used: {readable_format(used_ram)} | {percent_used_ram}%', 'red', attrs=['bold']))

	elif int(percent_used_ram) in range(91, 101):
		print(colored(f'Used: {readable_format(used_ram)} | {percent_used_ram}% | RAM USAGE IS TOO HIGH!!', 'magenta', attrs=['bold', 'blink']))

	if int(percent_free_ram) in range(0, 50):
		print(colored(f'Free: {free_ram} | {percent_free_ram - 100:.1f}%', 'green', attrs=['bold']))

	elif int(percent_free_ram) in range(51, 80):
		print(colored(f'Free: {free_ram} | {percent_free_ram - 100:.1f}%', 'yellow', attrs=['bold']))

	elif int(percent_free_ram) in range(81, 90):
		print(colored(f'Free: {free_ram} | {percent_free_ram - 100:.1f}%', 'red', attrs=['bold']))

	elif int(percent_free_ram) in range(91, 101):
		print(colored(f'Free: {free_ram} | {percent_free_ram - 100:.1f}% | LOW RAM MEMORY AVAILABLE', 'magenta', attrs=['bold', 'blink']))

	print(colored('|====== [Swap] =======|', 'green', attrs=['bold']))
	print(colored(f'Total: {total_swap}', 'blue', attrs=['bold']))

	if int(percent_used_swap) in range(0, 50):
		print(colored(f'Used: {used_swap} | {percent_used_swap}%', 'green', attrs=['bold']))

	elif int(percent_used_swap) in range(51, 80):
		print(colored(f'Used: {used_swap} | {percent_used_swap}%', 'yellow', attrs=['bold']))

	elif int(percent_used_swap) in range(81, 90):
		print(colored(f'Used: {used_swap} | {percent_used_swap}%', 'red', attrs=['bold']))

	elif int(percent_used_swap) in range(91, 101):
		print(colored(f'Used: {used_swap} | {percent_used_swap}% | SWAP USAGE IS TOO HIGH!!', 'magenta', attrs=['bold', 'blink']))

	if int(percent_free_swap) in range(0, 50):
		print(colored(f'Free: {free_swap} | {percent_free_swap - 100:.1f}%', 'green', attrs=['bold']))

	elif int(percent_free_swap) in range(51, 80):
		print(colored(f'Free: {free_swap} | {percent_free_swap - 100:.1f}%', 'yellow', attrs=['bold']))

	elif int(percent_free_swap) in range(81, 90):
		print(colored(f'Free: {free_swap} | {percent_free_swap - 100:.1f}%', 'red', attrs=['bold']))

	elif int(percent_free_swap) in range(91, 101):
		print(colored(f'Free: {free_swap} | {percent_free_swap - 100:.1f}% | LOW SWAP MEMORY AVAILABLE!!', 'magenta', attrs=['bold', 'blink']))

	print(colored('|======= [CPU] =======|', 'green', attrs=['bold']))

	cpu_perc = cpu_percent()

	if int(cpu_perc) in range(0, 50):
		print(colored(f'Usage: {cpu_perc}%', 'green', attrs=['bold']))

	elif int(cpu_perc) in range(51, 80):
		print(colored(f'Usage: {cpu_perc}%', 'yellow', attrs=['bold']))

	elif int(cpu_perc) in range(81, 90):
		print(colored(f'Usage: {cpu_perc}%', 'red', attrs=['bold']))

	elif int(cpu_perc) in range(91, 101):
		print(colored(f'Usage: {cpu_perc}% | RAM USAGE IS TOO HIGH!!', 'magenta', attrs=['bold', 'blink']))

	print(colored(f'Total cores: {cpu_count()}', attrs=['bold']))
	print(colored(f'Clusters: {CLUSTERS}', attrs=['bold']))
	print(colored(f'Cores per cluster: {CORES_PER_CLUSTER}', attrs=['bold']))
	print(colored(f'Threads per core: {THREADS_PER_CORE}', attrs=['bold']))

	corenum = 0

	for core in cpu:
		for ccore in range(1):
			corenum += 1
			
		print(colored(f'Core {corenum} frequency: {core[0]} GHz', attrs=['bold']))

	print(colored(f'Architecture type: {ARCH_TYPE}', attrs=['bold']))
	print(colored(f'CPU model/name: {CPU_MODEL}', attrs=['bold']))

	try:
		sleep(0.5)

	except KeyboardInterrupt:
		exit()

	system('clear')